src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs

```diff
@@ -66,10 +66,16 @@ public async Task SynchronizeWithBuildAsync(ImmutableDictionary<ProjectId, Immut
                     RaiseProjectDiagnosticsIfNeeded(project, stateSets, oldAnalysisData.Result, newResult);
                 }
 
-                // if we have updated errors, refresh open files
-                if (buildDiagnostics.Count > 0 && PreferLiveErrorsOnOpenedFiles(options))
+                // Refresh diagnostics for open files after solution build completes.
+                if (onBuildCompleted && PreferLiveErrorsOnOpenedFiles(options))
                 {
-                    // enqueue re-analysis of open documents.
+                    // Enqueue re-analysis of active document.
+                    if (_documentTrackingService?.GetActiveDocument(solution) is { } activeDocument)
+                    {
+                        AnalyzerService.Reanalyze(Workspace, documentIds: ImmutableArray.Create(activeDocument.Id), highPriority: true);
+                    }
+
+                    // Enqueue re-analysis of open documents.
                     AnalyzerService.Reanalyze(Workspace, documentIds: Workspace.GetOpenDocumentIds(), highPriority: true);
```

Comment:
aside (since you're not changing this), GetOpenDocuentsIds is always a weird API.  It's only valid for correctness when accessed on the UI thread, but it's valid for perf optimization no any thread.  Here, this seems to be used for correctness, but it opens us up to a race.  specifically, we might be on the BG when processing "build completed", but we analyze a file based on it's open-condition, even if that doesn't necessarily match if its' actually open.

not sure if it's an issue.

Reply:
Yes, I am not very sure either. This seems more like a perf case where we are just forcing higher priority analysis of open documents after build, instead of letting them get processed through regular normal priority analysis.